### PR TITLE
Improve course suggestions

### PR DESCRIPTION
**Pull request by LomyW**

### CODE REVIEW

**Review by Jonasdoubleyou — APPROVED** [review]

File: graphql/subcourse/fields.ts [hunk 1]
```diff
@@ -86,6 +87,12 @@ export class ExtendedFieldsSubcourseResolver {
                     { joinAfterStart: { equals: true }, lecture: { some: { start: { gt: new Date() } } } },
                 ],
             });
+            if (isSessionPupil(context)) {
+                const pupil = await getSessionPupil(context);
+                filters.push({
+                    AND: [{ minGrade: { lte: gradeAsInt(pupil.grade) }, maxGrade: { gte: gradeAsInt(pupil.grade) } }],
```
Comment (Jonasdoubleyou) [inline graphql/subcourse/fields.ts:93]: What if pupil.grade is null?

**Review by Jonasdoubleyou — CHANGES_REQUESTED** [review]

File: graphql/subcourse/fields.ts (hunk 1 above)
Comment (LomyW) [reply graphql/subcourse/fields.ts:93]: I think than the pupil shouldn't get a suggestion or? @Jonasdoubleyou

File: graphql/subcourse/fields.ts [hunk 2]
```diff
@@ -86,6 +87,16 @@ export class ExtendedFieldsSubcourseResolver {
                     { joinAfterStart: { equals: true }, lecture: { some: { start: { gt: new Date() } } } },
                 ],
             });
+            if (isSessionPupil(context)) {
+                const pupil = await getSessionPupil(context);
+                const pupilGrade = gradeAsInt(pupil.grade);
+                if (pupilGrade) {
+                    filters.push({
+                        AND: [{ minGrade: { lte: pupilGrade }, maxGrade: { gte: pupilGrade } }],
```
Comment (dhenkel92) [inline graphql/subcourse/fields.ts:95]: Out of curiosity, what is the intention of this PR?
It looks like we are trying to narrow down the courses based on the pupli who is requesting to show only the ones that are interesting to them.

But isn't this filter returning all the courses not available for his grade?

```suggestion
                        AND: [{ minGrade: { gte: pupilGrade }, maxGrade: { lte: pupilGrade } }],
```

File: graphql/subcourse/fields.ts [hunk 3]
```diff
@@ -86,6 +87,16 @@ export class ExtendedFieldsSubcourseResolver {
...
+                        // AND: [{ OR: [{ minGrade: { lte: pupilGrade } }, { minGrade: null }] }, { OR: [{ maxGrade: { gte: pupilGrade } }, { maxGrade: null }] }],
```
[the remark this answers is not in the document]
Comment (timidak) [reply graphql/subcourse/fields.ts]: Yes, sure. You want to get the subcourses where the min grade is lower or equal to the grade of the pupil and the max grade greater or equal. So the pupil must be in that range. 
E.g.:

```ts
pupilGrade = 8
sc1 = { minGrade: 10, maxGrade: 12 }
sc2 = { minGrade: 7, maxGrade: 10 } // <- it's a match for a suggestion
sc3 = { minGrade: 4, maxGrade: 7 }
sc4 = { minGrade: 8, maxGrade: 9 } // <- it's a match for a suggestion

```
Maybe the word `filters` confuses here a bit.

**Review by Jonasdoubleyou — APPROVED** [review]

**Pull request merged**